posts/2019-04-11-Rust-1.34.0.md

```diff
@@ -0,0 +1,222 @@
+with error values causing a nonzero exit code in the case of `fn main()`,
+and a test failure in the case of the tests.
+
+Support in `fn main()` and `#[test]` was implemented [many releases ago].
+However, the support within documentation tests was limited to doctests that have an explicit `fn main()`.
+
+In this release, full support for `?` in doctests has been added.
+Now, you can write this in your documentation tests:
+
+````rust
+/// ```rust
+/// use std::io;
+/// let mut input = String::new();
+/// io::stdin().read_line(&mut input)?;
+/// # Ok::<(), io:Error>(())
+/// ```
+fn my_func() {}
+````
+
+You still have to specify the error type being used at the bottom of the documentation test.
+
+### Custom attributes accept arbitrary token streams
+
+[Procedural macros]: https://blog.rust-lang.org/2018/12/21/Procedural-Macros-in-Rust-2018.html
+
+[Procedural macros] in Rust can define custom attributes that they consume.
+Until now, such attributes were restricted to being trees of paths and literals
+according to a specific syntax, like:
+
+```rust
+#[foo(bar)]
+#[foo = "bar"]
+#[foo = 0]
+#[foo(bar = true)]
+#[foo(bar, baz(quux, foo = "bar"))]
+```
+
+Unlike `macro_rules!`, these attributes could not accept arbitrary token streams in delimiters,
+so you could not write `#[range(0..10)]` or `#[bound(T: MyTrait)]`.
+Procedural macro crates would instead use strings for specifying syntaxes like this, e.g. `#[range("0..10")]`
+
+With this Rust release, custom attributes `#[attr($tokens)]` now accept
+arbitrary token streams in `$tokens`, bringing them on par with macros.
+If you're the author of a procedural macro crate, please check if your custom attributes
+have unnecessary strings in their syntax and if they can be better expressed with token streams.
+
+### `TryFrom` and `TryInto`
+
+[`from_be_bytes`]: https://doc.rust-lang.org/std/primitive.u32.html#method.from_be_bytes
+[never_type]: https://github.com/rust-lang/rust/issues/35121
+[`TryFrom`]: https://doc.rust-lang.org/std/convert/trait.TryFrom.html
+[`TryInto`]: https://doc.rust-lang.org/std/convert/trait.TryInto.html
+[`Infallible`]: https://doc.rust-lang.org/std/convert/enum.Infallible.html
+
+The [`TryFrom`] and [`TryInto`] traits were stabilized to allow fallible type conversions.
+
+For example, the [`from_be_bytes`] and related methods on integer types take arrays,
+but data is often read in via slices. Converting between slices and arrays is tedious to do manually.
+With the new traits, it can be done inline with `.try_into()`.
+
+```rust
+let num = u32::from_be_bytes(slice.try_into()?);
+```
+
+For conversions that can't fail, such as `u8` to `u32`, the [`Infallible`] type was added.
+This also permits a blanket implementation of `TryFrom` for all existing `From` implementations.
+In the future, we hope to turn `Infallible` into an alias for [the `!` (never) type][never_type].
+
+### `fn before_exec` deprecated in favor of `unsafe fn pre_exec`
```

Comment:
r? @RalfJung

Reply:
Might be worth mentioning the other concerns related to signal-safety in forked code?

> After a fork() in a multithreaded program, the child can safely call only async-signal-safe functions (see signal-safety(7)) until such time as it calls execve(2).

(From the fork man page)

For example, acquiring a lock in `pre_exec` is a very bad idea because if the lock was held by *someone* when the fork happened, now you got a deadlock.

Reply:
Sure, not sure how to phrase it. Could you file a PR against my branch or use GH suggestions?

Reply:
I'm afraid I don't have any good ideas either, and this POSIX stuff is not at all my expertise.

Reply:
[According to what the closest thing I know of to the POSIX spec is](http://pubs.opengroup.org/onlinepubs/007904975/functions/fork.html) it's similar to what @RalfJung quoted but subtly different:

> Consequently, to avoid errors, the child process may only execute async-signal-safe operations until such time as one of the exec functions is called

My interpretation that jives with my understanding of the API, which is that it's tough to use `before_exec` correctly but from the perspective of POSIX it's no UB, but rather we're explicitly saying it's UB with respect to file descriptor duplication (and how `FromRawFd` has various guarantees and whatnot).

I don't think we settled on any other concrete sequence of actions we'd consider UB other than fd trickery

Reply:
What I quoted is from `man fork` on my system, also [available online](https://www.systutorials.com/docs/linux/man/2-fork/). [This page](http://man7.org/linux/man-pages/man7/signal-safety.7.html) talks about "unpredictable results" and things being "unsafe to call" when in a not-async-signal-safe function in a signal handler; I do not know if that means UB or something weaker. I see no reason why calling them from a forked-off process should be any different.

Indeed we have no other UB that I know of, but I also did not try to construct a Rust library that is generally safe to use but UB when called in a not-async-signal-safe context. But we certainly have a chance of deadlocks (like when you try to acquire a mutex), and a case of nobody-knows-what-happens when you call into C code that does not declare that it is async-signal-safe.

Anyway my intention was not to say here that these things are UB, but to call out the fact that this is a constrained environment and very unpredictable but possibly not UB things can certainly happen.

Reply:
So I think we don't have time to add or change anything here unfortunately... Maybe this can be improved in the libstd docs later...